wrapper/src/main/java/software/amazon/jdbc/HikariPooledConnectionProvider.java

```diff
@@ -0,0 +1,127 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package software.amazon.jdbc;
+
+import com.zaxxer.hikari.HikariConfig;
+import com.zaxxer.hikari.HikariDataSource;
+import java.sql.Connection;
+import java.sql.SQLException;
+import java.util.List;
+import java.util.Map;
+import java.util.Properties;
+import java.util.concurrent.ConcurrentHashMap;
+import org.checkerframework.checker.nullness.qual.NonNull;
+import software.amazon.jdbc.cleanup.CanReleaseResources;
+import software.amazon.jdbc.util.HikariCPSQLException;
+import software.amazon.jdbc.util.RdsUrlType;
+import software.amazon.jdbc.util.RdsUtils;
+import software.amazon.jdbc.util.StringUtils;
+
+public abstract class HikariPooledConnectionProvider implements PooledConnectionProvider,
+    CanReleaseResources {
+
+  private static final RdsUtils rdsUtils = new RdsUtils();
+  private static final Map<String, HikariDataSource> databasePools = new ConcurrentHashMap<>();
+  private final HikariPoolConfigurator poolConfigurator;
+
+  public HikariPooledConnectionProvider(HikariPoolConfigurator hikariPoolConfigurator) {
+    poolConfigurator = hikariPoolConfigurator;
+  }
+
+  @Override
+  public boolean acceptsUrl(
+      @NonNull String protocol, @NonNull HostSpec hostSpec, @NonNull Properties props) {
+    final RdsUrlType urlType = rdsUtils.identifyRdsType(hostSpec.getHost());
+    return RdsUrlType.RDS_INSTANCE.equals(urlType);
+  }
+
+  @Override
+  public boolean acceptsStrategy(@NonNull HostRole role, @NonNull String strategy) {
+    return false;
+  }
+
+  @Override
+  public HostSpec getHostSpecByStrategy(
+      @NonNull List<HostSpec> hosts, @NonNull HostRole role, @NonNull String strategy) {
+    // This class does not accept any strategy, so the ConnectionProviderManager should prevent us
+    // from getting here.
+    return null;
+  }
+
+  @Override
+  public Connection connect(
+      @NonNull String protocol, @NonNull HostSpec hostSpec, @NonNull Properties props)
+      throws SQLException {
+    HikariDataSource ds = databasePools.computeIfAbsent(
+        hostSpec.getUrl(), url -> {
+          HikariConfig config = poolConfigurator.configurePool(hostSpec, props);
+          setConnectionProperties(config, hostSpec, props);
+          return new HikariDataSource(config);
+        });
+    return ds.getConnection();
+  }
+
+  @Override
+  public Connection connect(
+      @NonNull String url, @NonNull Properties props) throws SQLException {
+    // This method is only called by tests/benchmarks
+    return null;
+  }
+
+  @Override
+  public void releaseResources() {
+    databasePools.forEach((String url, HikariDataSource ds) -> ds.close());
+    databasePools.clear();
+  }
+
+  abstract String getDataSourceClassName();
+
+  private void setConnectionProperties(
```

Comment:
Probably protected would be better. There's a big chance that we might need to override it in subclasses. Just a suggestion.